manual/src/tutorials/effectexamples.etex

```diff
@@ -0,0 +1,666 @@
+threading systems provide "fork" primitive to spawn off a new concurrent task
+and a "yield" primitive to yield control to some other task. Correspondingly,
+we shall declare two effects as follows:
+
+\begin{caml_example*}{verbatim}
+type _ Effect.t += Fork : (unit -> unit) -> unit t
+                 | Yield : unit t
+\end{caml_example*}
+
+"Fork" effect takes a thunk and returns a unit to the performer. "Yield" effect
+is unparameterized, and returns a unit when performed. Let us consider that a
+task performing "Xchg" effect may match with any other task also offering to
+exchange a value.
+
+We shall also define helper functions that simply perform these effects:
+
+\begin{caml_example*}{verbatim}
+let fork f = perform (Fork f)
+let yield () = perform Yield
+let xchg v = perform (Xchg v)
+\end{caml_example*}
+
+A top-level "run" function defines the scheduler:
+
+\begin{caml_example*}{verbatim}
+(* A concurrent round-robin scheduler *)
+let run main =
+  let exchanger = ref None in (* waiting exchanger *)
+  let run_q = Queue.create () in (* scheduler queue *)
+  let enqueue k v =
+    Queue.push (fun _ -> continue k v) run_q
+  in
+  let dequeue () =
+    if Queue.is_empty run_q then () (* done *)
+    else (Queue.pop run_q) ()
+  in
+  let rec spawn f =
+    match_with f () {
+      retc = dequeue;
+      exnc = (fun e ->
+        print_endline (Printexc.to_string e);
+        dequeue ());
+      effc = fun (type a) (eff : a t) ->
+        match eff with
+        | Yield -> Some (fun (k : (a, unit) continuation) ->
+            enqueue k (); dequeue ())
+        | Fork f -> Some (fun (k : (a, unit) continuation) ->
+            enqueue k (); spawn f)
+        | Xchg n -> Some (fun (k : (int, unit) continuation) ->
+            begin match !exchanger with
+            | Some (n', k') ->
+                exchanger := None; enqueue k' n; continue k n'
+            | None -> exchanger := Some (n, k); dequeue ()
+            end)
+        | _ -> None
+    }
+  in
+  spawn main
+\end{caml_example*}
+
+We use a mutable queue "run_q" to hold the scheduler queue. "enqueue" pushes
+tasks into the queue, and "dequeue" pops tasks from the queue and runs them. We
+also have a ref cell "exchanger" that holds the task offering to exchange a
+value.
+
+The heavy lifting is done by the "spawn" function. The "spawn" function runs
+the given computation "f" in an effect handler. If "f" returns with a value
+(case "retc"), we dequeue and run the next task from the scheduler queue. If
+the computation "f" raises an exception (case "exnc"), we print the exception
+and run the next task from the scheduler.
+
+The computation "f" may also perform effects. If "f" performs the "Yield"
+effect, we enqueue the current task and run the next one from the scheduler
+queue. If the effect is "Fork f", we enqueue the current task, and recursively
+call "spawn" on "f", which run the given task in its own handler effectively
```

Comment:
"which run" should be "which runs". But, I don't think this part of the sentence is very clear. Mentioning "its own handler" suggests that several handlers are installed at the same time, which is not true. I would suggest simply explaining that the thread `k` is suspended (inserted into the queue of ready threads) while the thread `f` is executed immediately (via a tail call to `spawn f`). You might add that this is an arbitrary choice, and that one could also have chosen to execute `k` and  insert `f` into the ready queue.